sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/PortEvaluationStage.java

```diff
@@ -0,0 +1,125 @@
+package com.mesosphere.sdk.offer.evaluate;
+
+import com.mesosphere.sdk.offer.*;
+import org.apache.mesos.Protos;
+
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.IntStream;
+
+/**
+ * This class evaluates an offer for its port resources against an {@link OfferRequirement}, finding ports dynamically
+ * in the offer where specified by the framework and modifying {@link org.apache.mesos.Protos.TaskInfo} and
+ * {@link org.apache.mesos.Protos.ExecutorInfo} where appropriate so that the ports are available in their respective
+ * environments.
+ */
+public class PortEvaluationStage extends ResourceEvaluationStage implements OfferEvaluationStage {
+    private String portName;
+    private Integer port;
+
+    public PortEvaluationStage(Protos.Resource resource, String taskName, String portName, Integer port) {
+        super(resource, taskName);
+        this.portName = portName;
+        this.port = port;
+    }
+
+    public PortEvaluationStage(Protos.Resource resource, String portName, Integer port) {
+        this(resource, null, portName, port);
+    }
+
+    @Override
+    public void evaluate(
+            MesosResourcePool offerResourcePool,
+            OfferRequirement offerRequirement,
+            OfferRecommendationSlate offerRecommendationSlate) throws OfferEvaluationException {
+        // If this is from an existing pod with the dynamic port already assigned and reserved, just keep it.
+        Protos.CommandInfo commandInfo = getTaskName().isPresent() ?
+                offerRequirement.getTaskRequirement(getTaskName().get()).getTaskInfo().getCommand() :
+                offerRequirement.getExecutorRequirementOptional().get().getExecutorInfo().getCommand();
+        String taskPort = CommandUtils.getEnvVar(commandInfo, portName);
+        Integer assignedPort = port;
+        if (assignedPort == 0 && taskPort != null) {
+            assignedPort = Integer.parseInt(taskPort);
+        }
+
+        if (assignedPort == 0) {
+            // We don't want to dynamically consume a port that's explicitly claimed by this pod accidentally, so
```

Comment:
Could break this out into e.g. `private static int selectDynamicPort(MesosResourcePool, OfferRequirement) { ... }`